### PR TITLE
Allows using caches without having to configure them explicitly.

### LINKED ISSUE #2454: Allow using caches without having to configure them explicitly
Right now, Caffeine caches won't be created unless they are explicitly configured under `micronaut.caches.*`. It would be better to create them on demand with default values, and still let the users configure them if they want to customise the settings.

In addition to that, cache configurations should be toggleable (ie: allow to set `micronaut.caches.my-cache.enabled: true|false`

### PR DESCRIPTION
**Pull request by alvarosanchez**
Also allows caches to be enabled/disabled.

Fixes #2454

### CODE REVIEW

**Comment by jameskleeh** [comment]
What is the idea behind disabling a cache if you could still use it due to the dynamic cache feature? Even with `'micronaut.caches.test.enabled': false` I could still have `@Cacheable("test")` and it would work. Am I missing something?

**Comment by alvarosanchez** [comment]
By "it would work" you mean that it would use the cache? If that is the case, it was certainly not the intention.

**Comment by graemerocher** [comment]
Hmm shouldn't the dynamic cache manager return an implementation that doesn't cache if it is disabled?

**Comment by alvarosanchez** [comment]
Latest commit should address that corner case.

**Comment by alvarosanchez** [comment]
I'll refactor this to remove the enable/disable part, as it's more tricky than I anticipated.

Closing this PR in the meantime.

File: runtime/src/test/groovy/io/micronaut/cache/SyncCacheSpec.groovy [hunk 1]
```diff
@@ -258,8 +258,25 @@ class SyncCacheSpec extends Specification {
...
+
+        cleanup:
+        applicationContext.stop()
+    }
+
+    void "the dynamic cache manager can be disabled"() {
+        given:
+        ApplicationContext applicationContext = ApplicationContext.run(
+                'micronaut.caches.test.initialCapacity':1,
+                'micronaut.caches.dynamic': false
```
Comment (jameskleeh) [inline runtime/src/test/groovy/io/micronaut/cache/SyncCacheSpec.groovy]: Can you split this into its own test case?

**Comment by alvarosanchez** [comment]
> I think setting `micronaut.caches.dynamic: false` creates a sync cache bean with the name `dynamic`. Can you confirm?

Good catch! Fixed that
